File: functions/src/drives/route-thumbnail.ts
/**
 * Route thumbnails for the History list — a tiny, self-contained shape of the
 * drive that the list can draw with ZERO extra work.
 *
 * ## Why the geometry is stored on the ride document
 * The History list already reads `rides/{rideId}`. Anything the card needs that
 * is NOT on that document costs a per-card fetch: the full route lives in
 * member-gated Cloud Storage (`rideRoutes/{uid}/{rideId}/route.bin`) and can be
 * ~20 000 points, so drawing a shape per card from it would mean one Storage
 * download + decode per visible row in a scrolling list. Instead, drives.save
 * derives a ~64-point simplification ONCE, at save time, and stores it as an
 * encoded polyline (a few hundred bytes) on the document the list already has.
 * The card then costs zero reads, zero Storage fetches and zero map instances.
 *
 * ## This is coarse route geometry in Firestore — deliberately
 * The data model's rule that route GPS points live only in Cloud Storage still
 * governs the FULL track. This overview is a deliberate, bounded exception:
 * ~64 points at ~1 m encoding precision, enough to recognise the shape of a
 * drive and useless as a track. `rides/{rideId}` is owner-only read (plus
 * admin), exactly like every other field on it, so the thumbnail is visible to
 * nobody the distance and duration were not already visible to.
 *
 * ## Why Ramer–Douglas–Peucker and not every Nth point
 * Stride sampling keeps a fixed number of points but no particular ones: a
 * route whose only distinguishing feature is one sharp turn can lose the turn
 * entirely and render as a straight line. RDP keeps the points that carry the
 * SHAPE — it recursively retains whichever point is furthest from the chord
 * through a span and discards everything within a tolerance of it, so corners
 * survive and long straights collapse.
 *
 * Pure module — no Firebase Admin SDK imports.
 */

import type { TimedPoint } from './drive-calculations';

/**
 * Points kept in a stored thumbnail. Enough to keep a winding route legible at
 * the ~64x48 dp the card draws it at (well under a point per pixel of width),
 * small enough that the encoded string stays a few hundred bytes.
 */
export const THUMBNAIL_MAX_POINTS = 64;

/**
 * Coordinate precision of the stored polyline: 1e5, the classic Google
 * encoded-polyline scale (~1.1 m at the equator). Android's existing
 * `PolylineCodec` decodes with an explicit precision, so it reads this back
 * with no second decoder. 1e6 would double nothing useful here — the
 * simplification tolerance already discards far more than a metre.
 */
export const THUMBNAIL_POLYLINE_PRECISION = 1e5;

/** Metres per degree of latitude (mean spherical Earth) — the projection scale. */
const METRES_PER_DEGREE_LATITUDE = 111_320;

/**
 * Smallest simplification tolerance tried, in metres. Below this the encoded
 * string only grows: at the size the card draws, a metre is far under a pixel.
 */
const MIN_TOLERANCE_METRES = 5;

/**
 * Ceiling on the tolerance doubling. 5 m doubled 24 times is ~84 000 km — more
 * than twice the Earth's circumference, so the loop always terminates on the
 * point budget long before this. It exists only so a pathological input cannot
 * spin.
 */
const MAX_TOLERANCE_DOUBLINGS = 24;

/** A point projected to local metres (see {@link projectToMetres}). */
interface PlanarPoint {
  x: number;
  y: number;
}

/**
 * Equirectangular projection to local metres about the route's mean latitude.
 * Longitude degrees shrink by cos(latitude) — at Sweden's ~57-58°N a degree of
 * longitude is barely half a degree of latitude, so simplifying in raw degrees
 * would apply a tolerance twice as strict east-west as north-south and bias
 * which corners survive. Only distance RATIOS matter here, so a plane tangent
 * at the route's mean latitude is exact enough over any single drive.
 */
function projectToMetres(points: readonly TimedPoint[]): PlanarPoint[] {
  let latSum = 0;
  for (const point of points) latSum += point.latitude;
  const meanLatitude = latSum / points.length;
  const lonScale = Math.cos((meanLatitude * Math.PI) / 180);
  return points.map((point) => ({
    x: point.longitude * METRES_PER_DEGREE_LATITUDE * lonScale,
    y: point.latitude * METRES_PER_DEGREE_LATITUDE,
  }));
}

/** Perpendicular distance from `point` to the segment `start`→`end`. */
function perpendicularDistance(
  point: PlanarPoint,
  start: PlanarPoint,
  end: PlanarPoint,
): number {
  const dx = end.x - start.x;
  const dy = end.y - start.y;
  if (dx === 0 && dy === 0) {
    return Math.hypot(point.x - start.x, point.y - start.y);
  }
  // Cross product magnitude / segment length = perpendicular distance to the
  // INFINITE line. RDP's span endpoints are kept regardless, so the line form
  // is the right one: a point beyond an endpoint is still "far from the chord".
  return Math.abs(dy * point.x - dx * point.y + end.x * start.y - end.y * start.x) /
    Math.hypot(dx, dy);
}

/**
 * Indices kept by Ramer–Douglas–Peucker at `tolerance` metres, always including
 * the first and last point. Iterative (an explicit stack, not recursion) so a
 * 20 000-point track cannot blow the call stack.
 */
function rdpKeptIndices(points: readonly PlanarPoint[], tolerance: number): number[] {
  const keep = new Array<boolean>(points.length).fill(false);
  keep[0] = true;
  keep[points.length - 1] = true;

  const stack: Array<[number, number]> = [[0, points.length - 1]];
  while (stack.length > 0) {
    const [first, last] = stack.pop() as [number, number];
    let furthest = -1;
    let furthestDistance = tolerance;
    for (let i = first + 1; i < last; i += 1) {
      const distance = perpendicularDistance(
        points[i] as PlanarPoint,
        points[first] as PlanarPoint,
        points[last] as PlanarPoint,
      );
      if (distance > furthestDistance) {
        furthestDistance = distance;
        furthest = i;
      }
    }
    if (furthest !== -1) {
      keep[furthest] = true;
      stack.push([first, furthest], [furthest, last]);
    }
  }

  const kept: number[] = [];
  for (let i = 0; i < points.length; i += 1) {
    if (keep[i]) kept.push(i);
  }
  return kept;
}

/**
 * The route simplified to at most [maxPoints] points, preserving shape.
 *
 * RDP takes a tolerance, not a point count, so the tolerance is doubled from
 * [MIN_TOLERANCE_METRES] until the result fits the budget. Doubling (rather
 * than a binary search for the tightest fitting tolerance) is deliberate: it is
 * a handful of passes, it is deterministic, and the result is always a genuine
 * RDP simplification — never a truncation, which would drop the END of the
 * drive rather than its least interesting middle.
 *
 * Returns the input unchanged when it already fits.
 */
export function simplifyRoute(
  points: readonly TimedPoint[],
  maxPoints: number = THUMBNAIL_MAX_POINTS,
): TimedPoint[] {
  if (points.length <= maxPoints) return [...points];

  const planar = projectToMetres(points);
  let tolerance = MIN_TOLERANCE_METRES;
  let kept = rdpKeptIndices(planar, tolerance);
  for (let i = 0; i < MAX_TOLERANCE_DOUBLINGS && kept.length > maxPoints; i += 1) {
    tolerance *= 2;
    kept = rdpKeptIndices(planar, tolerance);
  }

  // A degenerate route (thousands of points at one spot) collapses to its two
  // endpoints and can never exceed the budget, so this is only a backstop.
  if (kept.length > maxPoints) {
    kept = kept.slice(0, maxPoints);
  }
  return kept.map((index) => points[index] as TimedPoint);
}

/** Encodes one signed, scaled coordinate delta in the polyline alphabet. */
function encodeSignedValue(value: number, out: string[]): void {
  let v = value < 0 ? ~(value << 1) : value << 1;
  while (v >= 0x20) {
    out.push(String.fromCharCode((0x20 | (v & 0x1f)) + 63));
    v >>= 5;
  }
  out.push(String.fromCharCode(v + 63));
}

/**
 * Encodes points as an encoded polyline at [THUMBNAIL_POLYLINE_PRECISION].
 *
 * This is the standard algorithm (zig-zag varint deltas, 5-bit chunks, +63 into
 * printable ASCII) — the same format Android's `PolylineCodec` already decodes
 * for Mapbox route geometry, so the app reads thumbnails back with the decoder
 * it has rather than a second, drifting one.
 */
export function encodePolyline(
  points: readonly { latitude: number; longitude: number }[],
  precision: number = THUMBNAIL_POLYLINE_PRECISION,
): string {
  const out: string[] = [];
  let previousLat = 0;
  let previousLon = 0;
  for (const point of points) {
    const lat = Math.round(point.latitude * precision);
    const lon = Math.round(point.longitude * precision);
    encodeSignedValue(lat - previousLat, out);
    encodeSignedValue(lon - previousLon, out);
    previousLat = lat;
    previousLon = lon;
  }
  return out.join('');
}

/**
 * The stored `routeThumbnail` for a submitted recording, or null when there is
 * no shape to draw.
 *
 * Null (not an empty string, and not a one-point polyline) for anything the
 * card cannot draw as a route: no points, a single fix, or a recording whose
 * fixes are all effectively the same place — a phone that sat parked with GPS
 * on. The card renders its placeholder for null, which is also what every drive
 * saved before this field existed gets, so the two paths are the same path.
 */
export function buildRouteThumbnail(points: readonly TimedPoint[] | undefined): string | null {
  if (!points || points.length < 2) return null;
  const simplified = simplifyRoute(points);
  if (simplified.length < 2) return null;
  // All points on one spot: after rounding to the encoding precision every
  // delta is zero, so the decoded shape is a dot. Say "no thumbnail" instead of
  // storing a polyline the card would have to special-case anyway.
  const first = simplified[0] as TimedPoint;
  const scale = THUMBNAIL_POLYLINE_PRECISION;
  const allSame = simplified.every(
    (point) =>
      Math.round(point.latitude * scale) === Math.round(first.latitude * scale) &&
      Math.round(point.longitude * scale) === Math.round(first.longitude * scale),
  );
  if (allSame) return null;
  return encodePolyline(simplified);
}

File: apps/android/app/src/main/java/com/kungsbackacarcommunity/app/drives/RouteThumbnail.kt
package com.kungsbackacarcommunity.app.drives

import com.kungsbackacarcommunity.app.navigation.LatLng
import com.kungsbackacarcommunity.app.navigation.PolylineCodec
import kotlin.math.abs
import kotlin.math.cos
import kotlin.math.max
import kotlin.math.min

/**
 * A point projected into a thumbnail's drawing box, in pixels from its top-left
 * corner. Deliberately NOT `androidx.compose.ui.geometry.Offset`: keeping the
 * projection's output a plain pair of floats is what lets the whole of
 * [RouteThumbnail] be unit-tested on the JVM with no Compose (and no device) in
 * the picture. The composable converts to a `Path` once, off the draw scope.
 */
data class ThumbnailPoint(val x: Float, val y: Float)

/**
 * Turns a drive's stored `routeThumbnail` — the ~64-point encoded polyline
 * `drives.save` derives from the recording — into a path the History card can
 * draw on a `Canvas`.
 *
 * ## Why this exists at all
 * A History card must cost nothing extra. The full route lives in member-gated
 * Cloud Storage and a card is a row in a scrolling list, so fetching or
 * decoding one per card, or instantiating a map per card, is out of the
 * question. The backend therefore puts a few hundred bytes of simplified
 * geometry on the ride document the list ALREADY reads, and this object turns
 * that string into ~64 points and a fitted box. No I/O, no map, no allocation
 * in the draw scope.
 *
 * ## Longitude is not latitude
 * A degree of longitude is a degree of latitude times cos(latitude). At
 * Kungsbacka's ~57.5°N that factor is ~0.54 — so projecting raw degrees into
 * the box would stretch every route to nearly twice its true width and turn a
 * square block of streets into a wide rectangle. The projection scales
 * longitude by cos(mean latitude) before fitting, which is exact enough over
 * the span of a single drive.
 *
 * ## Degenerate routes are the common case, not the edge
 * Every drive saved before the thumbnail existed has no polyline at all (there
 * is no backfill), and plenty of recordings are a phone that sat parked. Every
 * such case returns null from [project] and the card draws its placeholder —
 * one path, never an empty box and never a crash.
 *
 * Pure Kotlin (no Android, Compose or Firebase types) so all of it is unit
 * tested; only the `Canvas` call itself is device-only.
 */
object RouteThumbnail {

    /** Coordinate precision the backend encodes with (route-thumbnail.ts). */
    const val POLYLINE_PRECISION: Double = 1e5

    /** A polyline needs two points to be a line. */
    const val MIN_DRAWABLE_POINTS: Int = 2

    /**
     * Smallest bounding-box extent, in metres, still worth drawing as a route.
     *
     * Below this the fit would blow a few metres of GPS jitter up to fill the
     * whole box — a drive that never left the driveway would render as a
     * confident scribble that looks like a route and is not one. A stationary
     * recording says "no route overview" instead. 25 m is comfortably inside a
     * short street and comfortably outside a parked phone's wander.
     */
    const val MIN_EXTENT_METERS: Double = 25.0

    /** Metres per degree of latitude (mean spherical Earth). */
    private const val METERS_PER_DEGREE = 111_320.0

    /**
     * Coordinate bounds every real fix satisfies.
     *
     * A corrupt polyline that happens to decode cleanly still produces
     * coordinates: `"not a polyline!!"` decodes to a latitude of ~3623°. Those
     * are finite, so a finiteness check passes them, and the fit then scales
     * them into the box and draws a confident line that is not a route. The
     * backend only ever encodes points `drives.save` already validated to
     * ±90 / ±180, so anything outside is corruption by definition and gets the
     * placeholder — the same answer a stationary recording gets.
     */
    private const val MAX_ABS_LATITUDE = 90.0
    private const val MAX_ABS_LONGITUDE = 180.0

    /**
     * Decodes a stored thumbnail polyline, or an empty list when there is none
     * — including when the stored value is corrupt.
     *
     * Delegates to the app's existing [PolylineCodec] — the decoder already
     * shipped for Mapbox route geometry — rather than adding a second one that
     * could drift from the encoder. Only the precision differs (1e5 here, the
     * backend's; 1e6 for Mapbox), which the codec already takes as a parameter.
     *
     * ## Why the catch is here and not in the codec
     * [PolylineCodec] is not a total function and is not documented as one. A
     * polyline varint pair is latitude-then-longitude, so a string that ends
     * BETWEEN the two makes the decoder read past its last character and throw
     * `StringIndexOutOfBoundsException`. A single `"_"`, or one complete
     * latitude with nothing after it, is enough. A long run of garbage does
     * NOT do this — there is always another character left to consume, so it
     * merely decodes to nonsense — which is exactly why testing only a garbled
     * string never reached this path.
     *
     * The value arrives from Firestore and is decoded while composing a row in
     * the History LIST, so an escaping throw would not blank one card: it would
     * take down the whole History screen for anyone holding a single corrupt
     * document. "Corruption degrades to the placeholder" is this object's
     * stated contract, so this is the boundary that enforces it. Widening
     * [PolylineCodec] itself would instead change what its other caller sees —
     * Mapbox route geometry, where a malformed response is a real error worth
     * surfacing rather than silently drawing nothing.
     *
     * Catching [RuntimeException] rather than the narrower index error is
     * deliberate: the whole point of the boundary is that no decoder failure,
     * present or future, can reach the list.
     */
    fun decode(encoded: String?): List<LatLng> {
        if (encoded.isNullOrBlank()) return emptyList()
        return try {
            PolylineCodec.decode(encoded, POLYLINE_PRECISION)
        } catch (_: RuntimeException) {
            emptyList()
        }
    }

    /**
     * Fits [points] into a [width] x [height] box, inset by [padding] on every
     * side, preserving the route's true aspect ratio.
     *
     * Returns null — meaning "draw the placeholder" — when there is no route to
     * show: fewer than [MIN_DRAWABLE_POINTS] points, a box too small to draw
     * in, a non-finite or off-globe coordinate, or a route whose whole extent
     * is under [MIN_EXTENT_METERS].
     */
    fun project(
        points: List<LatLng>,
        width: Float,
        height: Float,
        padding: Float,
    ): List<ThumbnailPoint>? {
        if (points.size < MIN_DRAWABLE_POINTS) return null

        val innerWidth = width - 2 * padding
        val innerHeight = height - 2 * padding
        if (!innerWidth.isFinite() || !innerHeight.isFinite()) return null
        if (innerWidth <= 0f || innerHeight <= 0f) return null

        var minLat = Double.MAX_VALUE
        var maxLat = -Double.MAX_VALUE
        var minLon = Double.MAX_VALUE
        var maxLon = -Double.MAX_VALUE
        for (point in points) {
            if (!point.latitude.isFinite() || !point.longitude.isFinite()) return null
            if (abs(point.latitude) > MAX_ABS_LATITUDE) return null
            if (abs(point.longitude) > MAX_ABS_LONGITUDE) return null
            minLat = min(minLat, point.latitude)
            maxLat = max(maxLat, point.latitude)
            minLon = min(minLon, point.longitude)
            maxLon = max(maxLon, point.longitude)
        }

        // cos(mean latitude): the whole reason a Swedish route does not come
        // out ~1.9x too wide. Guarded against a degenerate pole value so the
        // scale can never be zero or negative.
        val lonScale = max(cos(Math.toRadians((minLat + maxLat) / 2.0)), 1e-6)

        // Extent in METRES, so the "is this actually a route" test means the
        // same thing everywhere on Earth rather than depending on latitude.
        val spanYMeters = (maxLat - minLat) * METERS_PER_DEGREE
        val spanXMeters = (maxLon - minLon) * METERS_PER_DEGREE * lonScale
        if (max(spanXMeters, spanYMeters) < MIN_EXTENT_METERS) return null

        // One scale for both axes = the route keeps its shape; the axis that
        // does not fill the box is centred in it. A perfectly straight
        // north-south (or east-west) route has zero span on one axis, which
        // must not become a division by zero — that axis simply does not
        // constrain the fit.
        val scaleX = if (spanXMeters > 0.0) innerWidth / spanXMeters else Double.MAX_VALUE
        val scaleY = if (spanYMeters > 0.0) innerHeight / spanYMeters else Double.MAX_VALUE
        val scale = min(scaleX, scaleY)
        if (!scale.isFinite() || scale <= 0.0) return null

        val drawnWidth = spanXMeters * scale
        val drawnHeight = spanYMeters * scale
        val originX = padding + (innerWidth - drawnWidth) / 2.0
        val originY = padding + (innerHeight - drawnHeight) / 2.0

        return points.map { point ->
            val xMeters = (point.longitude - minLon) * METERS_PER_DEGREE * lonScale
            // Latitude grows north, canvas y grows DOWN: measure from the top
            // (maxLat) so the drawn route is not upside down.
            val yMeters = (maxLat - point.latitude) * METERS_PER_DEGREE
            ThumbnailPoint(
                x = (originX + xMeters * scale).toFloat(),
                y = (originY + yMeters * scale).toFloat(),
            )
        }
    }

    /**
     * Convenience for the card: decode + project in one call, null when the
     * drive has no drawable thumbnail (including the very common case of a
     * drive saved before the field existed, where [encoded] is null).
     */
    fun pathFor(
        encoded: String?,
        width: Float,
        height: Float,
        padding: Float,
    ): List<ThumbnailPoint>? = project(decode(encoded), width, height, padding)
}
